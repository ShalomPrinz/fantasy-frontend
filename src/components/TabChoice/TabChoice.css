
.tab-choice {
    border: 4px solid white;
    border-style: solid;
}

.tab-choice:first-child {
    border-right: 3px solid white;
    border-top-left-radius: 1rem;
    border-bottom-left-radius: 1rem;
}

.tab-choice:last-child {
    border-left: 3px solid white;
    border-top-right-radius: 1rem;
    border-bottom-right-radius: 1rem;
}
