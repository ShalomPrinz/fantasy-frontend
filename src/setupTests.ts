// jest-dom adds custom jest matchers for asserting on DOM nodes.
// allows you to do things like:
// expect(element).toHaveTextContent(/react/i)
// learn more: https://github.com/testing-library/jest-dom
import '@testing-library/jest-dom';
import { ReactElement } from 'react';

import { render as renderToScreen, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';

const render = (Component: ReactElement) => renderToScreen(Component)

const setProperty = (obj: Object, prop: string, value: number) =>
    Object.defineProperty(obj, prop, 
        { writable: true, configurable: true, value: value })

const clickButton = (element: HTMLElement) => {
    userEvent.click(element)
}

const typeElement = (element: HTMLElement, input: string) =>
    userEvent.type(element, input)

export { clickButton, render, setProperty, screen, typeElement }
