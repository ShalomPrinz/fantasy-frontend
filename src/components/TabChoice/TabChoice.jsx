import React, { useState } from 'react';

import "./TabChoice.css"

const TabChoice = ({ tabs, defaultTab }) => {

    const [view, setView] = useState(defaultTab.Component);
    
    return (
        <>
            <div className="centered-flex py-4">
                {Object.entries(tabs).map(([name, props], index) => (
                    <button 
                        className="fs-2 bg-default py-2 px-5 tab-choice"
                        key={index}
                        onClick={() => {
                            setView(props.Component);
                            if (props.OnClick)
                                props.OnClick();
                        }}
                        type="button"
                    >
                        {name}
                    </button>
                ))} 
            </div>
            {view}
        </>
    );
}

export default TabChoice;
