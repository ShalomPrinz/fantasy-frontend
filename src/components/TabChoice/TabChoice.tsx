import { useState } from "react";

import { toast } from "react-toastify";

import { ConditionalList } from "../";

import "./TabChoice.css";

interface Tab {
  id: number;
  disabled?: {
    condition: () => boolean;
    toast: string;
  };
  label: string;
  Component: JSX.Element;
}

interface TabChoiceProps {
  tabs: Tab[];
}

const TabChoice = ({ tabs }: TabChoiceProps) => {
  const [activeTab, setActiveTab] = useState(tabs[0]);

  const tabCallback = (tab: Tab) => {
    const { id, disabled, label } = tab;

    const fontWeight = id === activeTab.id ? " fw-bold" : " fw-normal";
    const handleClick = () => {
      if (disabled?.condition()) toast.warn(disabled.toast);
      else setActiveTab(tab);
    };

    return (
      <button
        className={`fs-2 bg-default py-2 px-5 tab-choice${fontWeight}`}
        onClick={handleClick}
        type="button"
      >
        {label}
      </button>
    );
  };

  return (
    <>
      <div className="centered-flex py-4">
        <ConditionalList itemCallback={tabCallback} list={tabs} />
      </div>
      {activeTab?.Component}
    </>
  );
};

export default TabChoice;
