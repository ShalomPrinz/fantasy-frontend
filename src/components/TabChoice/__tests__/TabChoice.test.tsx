import { clickButton, render, screen } from 'setupTests'

import TabChoice from '../TabChoice';

interface Tab {
    id: number,
    label: string,
    Component: JSX.Element,
    onClick?: Function
}

describe('TabChoice', () => {
    it('should render empty div', () => {
        const tabs: Tab[] = []
        
        const { asFragment } = render(<TabChoice tabs={tabs} />);

        expect(asFragment()).toMatchSnapshot();
    })

    it('should render TabChoice component', () => {
        const tabs = [{ id: 0, label: 'One', Component: <h1>Hello</h1> }]
        
        const { asFragment } = render(<TabChoice tabs={tabs} />);

        expect(asFragment()).toMatchSnapshot();
    })
    
    describe('when tab is clicked', () => {
        it('should show the clicked tab component', () => {
            const tabs = [{ id: 0, label: 'One', Component: <h1>Hello</h1> }, { id: 1, label: 'Two', Component: <h1>World</h1> }]
        
            const { asFragment } = render(<TabChoice tabs={tabs} />);

            expect(asFragment()).toMatchSnapshot();
        
            const element = screen.getByText('Two')
            clickButton(element)
            expect(asFragment()).toMatchSnapshot();
        })

        it('should call given onClick prop', () => {
            const onClick = jest.fn()
            const tabs = [{ id: 0, label: 'One', Component: <h1>Hello</h1>, onClick: onClick }]

            render(<TabChoice tabs={tabs} />);
            const element = screen.getByText('One')
            clickButton(element)
            
            expect(onClick).toBeCalledTimes(1);
        })
    })
})
