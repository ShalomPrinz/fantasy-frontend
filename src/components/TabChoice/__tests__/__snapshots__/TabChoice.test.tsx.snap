// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TabChoice should render TabChoice component 1`] = `
<DocumentFragment>
  <div
    class="centered-flex py-4"
  >
    <button
      class="fs-2 bg-default py-2 px-5 tab-choice"
      type="button"
    >
      One
    </button>
  </div>
  <h1>
    Hello
  </h1>
</DocumentFragment>
`;

exports[`TabChoice should render empty div 1`] = `
<DocumentFragment>
  <div
    class="centered-flex py-4"
  />
</DocumentFragment>
`;

exports[`TabChoice when tab is clicked should show the clicked tab component 1`] = `
<DocumentFragment>
  <div
    class="centered-flex py-4"
  >
    <button
      class="fs-2 bg-default py-2 px-5 tab-choice"
      type="button"
    >
      One
    </button>
    <button
      class="fs-2 bg-default py-2 px-5 tab-choice"
      type="button"
    >
      Two
    </button>
  </div>
  <h1>
    Hello
  </h1>
</DocumentFragment>
`;

exports[`TabChoice when tab is clicked should show the clicked tab component 2`] = `
<DocumentFragment>
  <div
    class="centered-flex py-4"
  >
    <button
      class="fs-2 bg-default py-2 px-5 tab-choice"
      type="button"
    >
      One
    </button>
    <button
      class="fs-2 bg-default py-2 px-5 tab-choice"
      type="button"
    >
      Two
    </button>
  </div>
  <h1>
    World
  </h1>
</DocumentFragment>
`;
